### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by shuolinl**
In this pr:
1.  fix list output. when we list user/role's privileges, we will get: rolename, path, privilege,grantOption
2. git a func to check whether a user have role.
3. fix regrant. duplicate grant operation will not get error.

### CODE REVIEW

**Comment by shuolinl** [comment]
截屏2023-09-06 09 11 49
最终的效果。

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/auth/AuthorityChecker.java [hunk 1]
```diff
@@ -187,32 +194,89 @@ public static boolean checkGrantOption(
...
 
-    for (String header : authorizerInfo.keySet()) {
-      headerList.add(new ColumnHeader(header, TSDataType.TEXT));
+    if (listRoleUser) {
+      headerList.add(new ColumnHeader(authResp.getTag(), TSDataType.TEXT));
+    } else {
+      headerList.add(new ColumnHeader(new String("  ROLE  "), TSDataType.TEXT));
+      headerList.add(new ColumnHeader(new String("  PATH  "), TSDataType.TEXT));
+      headerList.add(new ColumnHeader(new String("  PRIVILEGES  "), TSDataType.TEXT));
+      headerList.add(new ColumnHeader(new String("  GRANT OPTION  "), TSDataType.TEXT));
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/auth/AuthorityChecker.java:228]:
```suggestion
      headerList.add(new ColumnHeader(new String("ROLE"), TSDataType.TEXT));
      headerList.add(new ColumnHeader(new String("PATH"), TSDataType.TEXT));
      headerList.add(new ColumnHeader(new String("PRIVILEGES"), TSDataType.TEXT));
      headerList.add(new ColumnHeader(new String("GRANT OPTION"), TSDataType.BOOLEAN));
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/auth/AuthorityChecker.java [hunk 2]
```diff
@@ -187,32 +194,89 @@ public static boolean checkGrantOption(
...
   }
+
+  private static void appendPriBuilder(
+      String name, String path, Set<Integer> priv, Set<Integer> grantOpt, TsBlockBuilder builder) {
+    for (int i : priv) {
+      builder.getColumnBuilder(0).writeBinary(new Binary(new String(name)));
+      builder.getColumnBuilder(1).writeBinary(new Binary(new String(path)));
+      builder.getColumnBuilder(2).writeBinary(new Binary(PrivilegeType.values()[i].toString()));
+      if (grantOpt.contains(i)) {
+        builder.getColumnBuilder(3).writeBinary(new Binary(new String("T")));
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/auth/AuthorityChecker.java:274]:
```suggestion
        builder.getColumnBuilder(3).writeBoolean(true);
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/auth/AuthorityChecker.java [hunk 3]
```diff
@@ -187,32 +194,89 @@ public static boolean checkGrantOption(
...
+      } else {
+        builder.getColumnBuilder(3).writeBinary(new Binary(new String("F")));
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/auth/AuthorityChecker.java:276]:
```suggestion
        builder.getColumnBuilder(3).writeBoolean(false);
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/AuthorInfo.java [hunk 4]
```diff
@@ -515,6 +465,23 @@ public TPermissionInfoResp checkUserPrivilegeGrantOpt(
...
+    if (user == null) {
+      throw new AuthException(
+          TSStatusCode.USER_NOT_EXIST, String.format("No such user : %s", username));
+    }
+    result = getUserPermissionInfo(username);
+    if (!user.getRoleList().contains(rolename)) {
+      result.setStatus(RpcUtils.getStatus(TSStatusCode.SUCCESS_STATUS));
+    } else {
+      result.setStatus(RpcUtils.getStatus(TSStatusCode.USER_NOT_HAS_ROLE));
+    }
```
Comment (JackieTien97) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/AuthorInfo.java:481]: reserse it

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**